### ISSUE TITLE
npm install sqlite3 --build-from-source (fails)

### ISSUE THREAD

**Issue by fractalf** [mapbox/node-sqlite3#459]
I've had heaps of problems with sqlite3 earlier, but mostly managed to figure it out by reading the errors carefully. This time however I'm quite stuck. Any ideas?

```
FROM alpine:edge

RUN echo "http://nl.alpinelinux.org/alpine/edge/main" > /etc/apk/repositories && \
    echo "http://nl.alpinelinux.org/alpine/edge/testing" >> /etc/apk/repositories && \
    apk add -U iojs wget ca-certificates python make gcc libc-dev g++ && \
    apk upgrade && \
    npm install sqlite3 --build-from-source
```

this will give error

```
...
../src/statement.cc: In static member function 'static void node_sqlite3::Statement::Work_BeginReset(node_sqlite3::Statement::Baton*)':
../src/macros.h:122:9: warning: unused variable 'status' [-Wunused-variable]
     int status = uv_queue_work(uv_default_loop(),                              \
         ^
../src/statement.cc:725:5: note: in expansion of macro 'STATEMENT_BEGIN'
     STATEMENT_BEGIN(Reset);
     ^
  SOLINK_MODULE(target) Release/obj.target/node_sqlite3.node
/usr/lib/gcc/x86_64-alpine-linux-musl/5.1.0/../../../../x86_64-alpine-linux-musl/bin/ld: Release/obj.target/node_sqlite3.node: No symbol version section for versioned symbol `memcpy@GLIBC_2.2.5'
/usr/lib/gcc/x86_64-alpine-linux-musl/5.1.0/../../../../x86_64-alpine-linux-musl/bin/ld: final link failed: Nonrepresentable section on output
collect2: error: ld returned 1 exit status
node_sqlite3.target.mk:130: recipe for target 'Release/obj.target/node_sqlite3.node' failed
make: *** [Release/obj.target/node_sqlite3.node] Error 1
make: Leaving directory '/node_modules/sqlite3/build'
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:269:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:199:12)
gyp ERR! System Linux 3.13.0-43-generic
gyp ERR! command "/usr/bin/iojs" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "build" "--fallback-to-build" "--module=/node_modules/sqlite3/lib/binding/node-v44-linux-x64/node_sqlite3.node" "--module_name=node_sqlite3" "--module_path=/node_modules/sqlite3/lib/binding/node-v44-linux-x64"
gyp ERR! cwd /node_modules/sqlite3
gyp ERR! node -v v2.2.1
gyp ERR! node-gyp -v v1.0.3
gyp ERR! not ok
...
```

I suspect it has to do with memcpy@GLIBC_2.2.5, but really I couldn't find out what.

Full log (commands and all errors)
http://pastebin.com/YZShJzY2

Thanks for any input

**Comment by springmeyer** [comment]
Looks like you are on alpine linux which uses http://www.musl-libc.org/ for the C library. Sorry this is broken for you - would love to make sure musl works, but don't have an easy way to test it.

My hunch, like you is it is the glibc hack that is breaking.

Can you try downloading node-sqlite3 from github master and apply this patch:

```diff
diff --git a/src/gcc-preinclude.h b/src/gcc-preinclude.h
index 38c9138..538cf64 100644
--- a/src/gcc-preinclude.h
+++ b/src/gcc-preinclude.h
@@ -1,6 +1,6 @@
 
 // https://rjpower9000.wordpress.com/2012/04/09/fun-with-shared-libraries-version-glibc_2-14-not-found/
 
-#if defined(__linux__) && defined(__x86_64__)
+#if defined(__GLIBC__) && defined(__linux__) && defined(__x86_64__)
 __asm__(".symver memcpy,memcpy@GLIBC_2.2.5");
 #endif
```

I'm hoping that `musl` does not define `__GLIBC__`, but I'm open to other suggestions for a way to detect glibc /cc @bnoordhuis

**Comment by fractalf** [comment]
I downloaded the github master files. When doing `./configure`i get error

`
gyp: Undefined variable module_name in binding.gyp while trying to load binding.gyp
`

Tried to search this, but all I found was others having the same problem, and could see a way out of it. So I'm stuck without being able to test the patch you gave me.

..worst case, this probably workes itself out in some updates in the future

**Comment by bmoelk** [comment]
I'm in the same boat. I was able to build the master branch after applying the patch:

    # apk-install gcc make libc-dev python g++
    # npm install
    # npm install -g node-pre-gyp
    # node-pre-gyp configure
    # node-pre-gyp build verbose

This seemed to resolve my immediate issue after using the freshly built build/Release/node_sqlite3.node

**Comment by springmeyer** [comment]
thanks for confirming @bmoelk - fixed pushed in 55db374. Will roll v3.0.9 shortly that includes this.

**Comment by fractalf** [comment]
Thanks @bmoelk !

**Comment by springmeyer** [comment]
sorry, going to revert the above fix for alpine linux since it turns out it disabled the original fix for older centos/etc. - refs https://github.com/mapbox/node-sqlite3/issues/484

**Comment by springmeyer** [comment]
TODO:

  - [ ] figure out why `__GLIBC__` appears not to be defined for compilers on travis (that build the binaries)
  - [ ] figure out a robust way to avoid using the symbol versioning hack for musl libc (and other libc's that are not the system glibc)

**Comment by s-a** [comment]
ref: https://github.com/s-a/iron-node/issues/42

**Comment by farcaller** [comment]
Any plan to address broken alpine builds?

**Comment by motet-a** [comment]
Can someone explain me why the [gcc-preclude.h] hack is still present? On old CentOS versions, it should be possible to rebuild anything with an old glibc without any problem (at least with `npm install --build-from-source`).

These regressions could be automatically tested with Docker.

Thanks.

[gcc-preclude.h]: https://github.com/mapbox/node-sqlite3/blob/master/src/gcc-preinclude.h

**Comment by TooTallNate** [comment]
Getting bit by this one as well. Any updates?

**Comment by springmeyer** [comment]
Can someone help me replicate? Just tried https://github.com/mapbox/node-sqlite3/issues/459#issuecomment-218282156 on OS X with:

```
docker -v
Docker version 17.06.2-ce, build cec0b72
```

And seeing:

```
docker run --rm -ti mhart/alpine-node 'apk --update add gcc g++ make python && npm install sqlite3'
Unable to find image 'mhart/alpine-node:latest' locally
latest: Pulling from mhart/alpine-node
88286f41530e: Pull complete 
94250b0c5a4e: Pull complete 
Digest: sha256:d0e18e64a06830d1e0a8fa7895c245385c6aeb323803b1aa97faded0c9cb4030
Status: Downloaded newer image for mhart/alpine-node:latest
docker: Error response from daemon: oci runtime error: container_linux.go:262: starting container process caused "exec: \"apk --update add gcc g++ make python && npm install sqlite3\": executable file not found in $PATH".
```

**Comment by TooTallNate** [comment]
@springmeyer Drop the quotes:

```
docker run --rm -ti mhart/alpine-node apk --update add gcc g++ make python && npm install sqlite3
```

**Comment by springmeyer** [comment]
@TooTallNate thanks. Will try that. In the meantime I've queued up a fix that I think should work on alpine: https://github.com/mapbox/node-sqlite3/pull/875

Can you try that branch?

**Comment by springmeyer** [comment]
Curiously the testcase works for me without errors. I would have expected the binaries to fallback to a source compile (given incompatible musl lib) but they install from remote and work fine. Is this different than what you are seeing @TooTallNate ?

```
docker run --rm -ti mhart/alpine-node
/ # apk --update add gcc g++ make python && npm install sqlite3
fetch http://dl-cdn.alpinelinux.org/alpine/v3.6/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.6/community/x86_64/APKINDEX.tar.gz
(1/24) Installing binutils-libs (2.28-r2)
(2/24) Installing binutils (2.28-r2)
(3/24) Installing gmp (6.1.2-r0)
(4/24) Installing isl (0.17.1-r0)
(5/24) Installing libgomp (6.3.0-r4)
(6/24) Installing libatomic (6.3.0-r4)
(7/24) Installing pkgconf (1.3.7-r0)
(8/24) Installing mpfr3 (3.1.5-r0)
(9/24) Installing mpc1 (1.0.3-r0)
(10/24) Installing gcc (6.3.0-r4)
(11/24) Installing musl-dev (1.1.16-r13)
(12/24) Installing libc-dev (0.7.1-r0)
(13/24) Installing g++ (6.3.0-r4)
(14/24) Installing make (4.2.1-r0)
(15/24) Installing libbz2 (1.0.6-r5)
(16/24) Installing expat (2.2.0-r1)
(17/24) Installing libffi (3.2.1-r3)
(18/24) Installing gdbm (1.12-r0)
(19/24) Installing ncurses-terminfo-base (6.0-r8)
(20/24) Installing ncurses-terminfo (6.0-r8)
(21/24) Installing ncurses-libs (6.0-r8)
(22/24) Installing readline (6.3.008-r5)
(23/24) Installing sqlite-libs (3.18.0-r0)
(24/24) Installing python2 (2.7.13-r1)
Executing busybox-1.26.2-r5.trigger
OK: 205 MiB in 37 packages

> sqlite3@3.1.10 install /node_modules/sqlite3
> node-pre-gyp install --fallback-to-build

[sqlite3] Success: "/node_modules/sqlite3/lib/binding/node-v57-linux-x64/node_sqlite3.node" is installed via remote
npm WARN saveError ENOENT: no such file or directory, open '/package.json'
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN enoent ENOENT: no such file or directory, open '/package.json'
npm WARN !invalid#1 No description
npm WARN !invalid#1 No repository field.
npm WARN !invalid#1 No README data
npm WARN !invalid#1 No license field.

+ sqlite3@3.1.10
added 137 packages in 7.864s
/ # ls node_modules/
nan      sqlite3
/ # node -e  "require('sqlite3')"
/ # node -e  "console.log(require('sqlite3'))"
{ Database: [Function: Database],
  Statement: [Function: Statement],
  OPEN_READONLY: 1,
  OPEN_READWRITE: 2,
  OPEN_CREATE: 4,
  VERSION: '3.15.0',
  SOURCE_ID: '2016-10-14 10:20:30 707875582fcba352b4906a595ad89198d84711d8',
  VERSION_NUMBER: 3015000,
  OK: 0,
  ERROR: 1,
  INTERNAL: 2,
  PERM: 3,
  ABORT: 4,
  BUSY: 5,
  LOCKED: 6,
  NOMEM: 7,
  READONLY: 8,
  INTERRUPT: 9,
  IOERR: 10,
  CORRUPT: 11,
  NOTFOUND: 12,
  FULL: 13,
  CANTOPEN: 14,
  PROTOCOL: 15,
  EMPTY: 16,
  SCHEMA: 17,
  TOOBIG: 18,
  CONSTRAINT: 19,
  MISMATCH: 20,
  MISUSE: 21,
  NOLFS: 22,
  AUTH: 23,
  FORMAT: 24,
  RANGE: 25,
  NOTADB: 26,
  cached: { Database: [Function: Database], objects: {} },
  verbose: [Function] }
```

**Comment by TooTallNate** [comment]
@springmeyer Yes it's different because you're also missing the `--build-from-source` flag.

Also your `memcpy-on-travis` branch works 👍 for me

**Comment by springmeyer** [comment]
> Also your memcpy-on-travis branch works 👍 for me

That is great.

> @springmeyer Yes it's different because you're also missing the --build-from-source flag.

Thanks for clarifying - I figured you were hitting that per my PR, but was trying to faithfully replicate too. I had assumed the binaries were not working. Otherwise I'm unsure why anyone would need to use `--build-from-source`

**Comment by TooTallNate** [comment]
> Otherwise I'm unsure why anyone would need to use `--build-from-source`

I'll explain it over 🍻 sometime 😄

**Comment by springmeyer** [comment]
sqlite3@3.1.11 is now out with the fix from #875

> I'll explain it over 🍻 sometime 😄

Sounds good :)
